internal/pkg/db/mongo/metadata.go

```diff
@@ -455,131 +455,156 @@ func (m MongoClient) GetDevice(d *contract.Device, q bson.M) error {
-	} else {
-		err := errors.New("mgoGetDeviceProfileById Invalid Object ID " + id)
-		return err
+		return m.getDeviceProfile(bson.M{"_id": bson.ObjectIdHex(id)})
+	}
+	if _, err := uuid.Parse(id); err != nil {
+		return contract.DeviceProfile{}, db.ErrInvalidObjectId
 	}
+	return m.getDeviceProfile(bson.M{"uuid": id})
 }
 
-func (m MongoClient) GetAllDeviceProfiles(dp *[]contract.DeviceProfile) error {
-	return m.GetDeviceProfiles(dp, nil)
+func (m MongoClient) GetAllDeviceProfiles() ([]contract.DeviceProfile, error) {
+	return m.getDeviceProfiles(nil)
 }
 
-func (m MongoClient) GetDeviceProfilesByModel(dp *[]contract.DeviceProfile, model string) error {
-	return m.GetDeviceProfiles(dp, bson.M{"model": model})
+func (m MongoClient) GetDeviceProfilesByModel(model string) ([]contract.DeviceProfile, error) {
+	return m.getDeviceProfiles(bson.M{"model": model})
 }
 
-func (m MongoClient) GetDeviceProfilesWithLabel(dp *[]contract.DeviceProfile, l string) error {
+func (m MongoClient) GetDeviceProfilesWithLabel(l string) ([]contract.DeviceProfile, error) {
 	var ls []string
 	ls = append(ls, l)
-	return m.GetDeviceProfiles(dp, bson.M{"labels": bson.M{"$in": ls}})
+
+	return m.getDeviceProfiles(bson.M{"labels": bson.M{"$in": ls}})
 }
-func (m MongoClient) GetDeviceProfilesByManufacturerModel(dp *[]contract.DeviceProfile, man string, mod string) error {
-	return m.GetDeviceProfiles(dp, bson.M{"manufacturer": man, "model": mod})
+
+func (m MongoClient) GetDeviceProfilesByManufacturerModel(man string, mod string) ([]contract.DeviceProfile, error) {
+	return m.getDeviceProfiles(bson.M{"manufacturer": man, "model": mod})
 }
-func (m MongoClient) GetDeviceProfilesByManufacturer(dp *[]contract.DeviceProfile, man string) error {
-	return m.GetDeviceProfiles(dp, bson.M{"manufacturer": man})
+
+func (m MongoClient) GetDeviceProfilesByManufacturer(man string) ([]contract.DeviceProfile, error) {
+	return m.getDeviceProfiles(bson.M{"manufacturer": man})
 }
-func (m MongoClient) GetDeviceProfileByName(dp *contract.DeviceProfile, n string) error {
-	return m.GetDeviceProfile(dp, bson.M{"name": n})
+
+func (m MongoClient) GetDeviceProfileByName(n string) (contract.DeviceProfile, error) {
+	return m.getDeviceProfile(bson.M{"name": n})
 }
 
 // Get device profiles with the passed query
-func (m MongoClient) GetDeviceProfiles(d *[]contract.DeviceProfile, q bson.M) error {
+func (m MongoClient) getDeviceProfiles(q bson.M) (cdps []contract.DeviceProfile, err error) {
 	s := m.session.Copy()
 	defer s.Close()
 	col := s.DB(m.database.Name).C(db.DeviceProfile)
 
 	// Handle the DBRefs
-	var mdps []mongoDeviceProfile
-	err := col.Find(q).Sort("queryts").All(&mdps)
+	var dps []models.DeviceProfile
+	err = col.Find(q).Sort("queryts").All(&dps)
 	if err != nil {
-		return err
+		return []contract.DeviceProfile{}, err
 	}
 
-	*d = []contract.DeviceProfile{}
-	for _, mdp := range mdps {
-		*d = append(*d, mdp.DeviceProfile)
+	for _, dp := range dps {
```

Comment:
This bit me earlier this week.  You need to initialize `cdps` to an empty slice here.  It defaults to nil.  Which will be returned if `dps` is empty.